src/dan.js

```diff
@@ -0,0 +1,337 @@
+        if (params['on_register'] != undefined && params['register_callback'] != undefined) {
+            throw new RegistrationError(_reg_msg);
+        }
+        else if (params['on_register'] != undefined) {
+            this.ctx.on_register = params['on_register'];
+        }
+        else if (params['register_callback'] != undefined) {
+            console.warning(_reg_msg);
+            this.ctx.on_register = params['register_callback'];
+        }
+
+        // other callbacks
+        this.ctx.on_deregister = params['on_deregister'];
+        this.ctx.on_connect = params['on_connect'];
+        this.ctx.on_disconnect = params['on_disconnect'];
+
+
+        // filter out the empty `df_list`, in case of empty list, server reponsed 403.
+        ['idf_list', 'odf_list'].forEach(
+            x => {
+                if (Array.isArray(body[x]) && body[x].length == 0)
+                    delete body[x];
+            }
+        );
+
+        new Promise(
+            (resolve, reject) => {
+                superagent.put(`${this.ctx.url}/${this.ctx.app_id}`)
+                    .type('json')
+                    .accept('json')
+                    .send(body)
+                    .then(res => {
+                        resolve(res);
+                    }, err => {
+                        reject(err);
+                    });
+            })
+            .then(res => {
+                let metadata = res.body;
+                if (typeof metadata === 'string') {
+                    metadata = JSON.parse(metadata);
+                }
+
+                this.ctx.name = metadata['name'];
+                this.ctx.mqtt_host = metadata['url']['host'];
+                this.ctx.mqtt_port = metadata['url']['ws_port'];
+                this.ctx.mqtt_username = metadata['username'] || '';
+                this.ctx.mqtt_password = metadata['password'] || '';
+                this.ctx.i_chans['ctrl'] = metadata['ctrl_chans'][0];
+                this.ctx.o_chans['ctrl'] = metadata['ctrl_chans'][1];
+                this.ctx.rev = metadata['rev'];
+
+                this.ctx.mqtt_client = mqtt.connect(`${metadata.url['ws_scheme']}://${this.ctx.mqtt_host}:${this.ctx.mqtt_port}`, {
+                    clientId: `iottalk-js-${this.ctx.app_id}`,
+                    username: this.ctx.mqtt_username,
+                    password: this.ctx.mqtt_password,
+                    will: {
+                        topic: this.ctx.i_chans['ctrl'],
+                        // in most case of js DA, it never connect back
+                        payload: JSON.stringify({ 'state': 'offline', 'rev': this.ctx.rev }),
+                        retain: true,
+                    },
+                    keepalive: 30,  // seems 60 is problematic for default mosquitto setup
+                });
+
+                this.ctx.mqtt_client.on('connect', (connack) => {
+                    console.info('mqtt_connect');
+                    this.on_connect();
```

Comment:
那麼 console.info 也放到 this.on_connect 裡面或拿掉吧。
這裡 `mqtt_client.on` 的 callback 就不需要多一個 anonymous function 了

Reply:
我嘗試改成這樣的形式會出現 
`TypeError: The "listener" argument must be of type Function. Received type undefine`
```suggestion
                this.ctx.mqtt_client.on('connect', this.on_connect());
```

Reply:
理論上是傳 function pointer，而不是執行結果
```
                this.ctx.mqtt_client.on('connect', this.on_connect);
```